### PR TITLE
Improvements to the Describe >Two\Three Variable > Summaries 

### PR DESCRIPTION
**Pull request by Ivanluv**
This PR has Skimr additions from PR #7267 and additional changes for the customised option i.e 

- Having the categorical by categorical option print frequency tables .

- Also there's an additional optional receiver  numerical that will have a summary table printed
This changes however only work correctly when we have one factor selected in the multiple receiver

### CODE REVIEW

**Comment by rdstern** [comment]
@Ivanluv there is quite a long way to go here.  
a) When it is factor by factor it is correctly a frequency table.  But the idea of your new system is that you can give all the options of counts or percentages.  And you can include margins.  So all these just as they are in the general frequencies in your tables stuff.
b) When the numeric variable is added the type of summary (on the left-hand side, should change to a summary table.  And it should be a summary table - you ignore the numeric variable and it is still a frequency table.  And the margins options still apply.
Separately from this I have realised there is another option for a set of summary tables.  The multiple receiver could be numeric, and then the two single receivers could both be factors.  That would give a set of summary tables.  

I think it is also time to involve @lilyclements in the coding for the different options.

**Comment by rdstern** [comment]
Sorry @Ivanluv I now have a slightly different suggestion.  Check first that you like it too and I'll also ask @lilyclements 
By the way, at the top change the title from Describe Two Variables to Summarise Two/three Variables.

I think this might get too complicated, so suggest a 3rd radio button at the top

So it is Skim, Two Variables, Three Variables.

If you agree then the first job is to complete the 2 variables, and that is 
a) the percentages, margins and other options for the frequency tables.  Checkl on the gerneral frequency table

b) Change the

**Comment by rdstern** [comment]
Ooops didn't mean to close!

**Comment by Ivanluv** [comment]
> Sorry @Ivanluv I now have a slightly different suggestion. Check first that you like it too and I'll also ask @lilyclements By the way, at the top change the title from Describe Two Variables to Summarise Two/three Variables.
> 
> I think this might get too complicated, so suggest a 3rd radio button at the top
> 
> So it is Skim, Two Variables, Three Variables.
> 
> If you agree then the first job is to complete the 2 variables, and that is a) the percentages, margins and other options for the frequency tables. Checkl on the gerneral frequency table
> 
> b) Change the

I think this is a good idea

**Comment by Ivanluv** [comment]
Am still making changes on this

**Comment by rdstern** [comment]
@Ivanluv ok, tell me when I can test again.
I am now assuming you will add the third button for the 3-way, and then concentrate on getting the two-way options working first.  So the extra control for the 2-way summaries is now to be part of the 3-way.

**Comment by Ivanluv** [comment]
@rdstern  for now you can test the two variable functionality ie the categorical by categorical option

**Comment by rdstern** [comment]
I looked at the frequency by frequency and it is looking good. 
a) The third variable needs to be moved to the new three variable button - and it is then not optional, it is called Third Variable.  It might stay there for now, because the dialogue will look so similar.
b) Perhaps swap the percentage box and the display box.
c) In the percentage box here, this should change to a single receiver.  It is multiple in the 3 way.  Currently it allows all factors, even those that are not in the table.  Can you restrict it?  If not, then it is ignored, so not as problem.
d) Put the Display Outer into the Display group box.  Do you really want to call it Outer?  I would prefer Margin.  
e)  You have the Column Factors as numerical and it doesn't seem to do anything here.  I wonder what you want, and you could also look forward to having 3 variables in a similar-looking control.  One possibility would be to just allow 1 or 2 here, so the default is 2, i.e. the 2nd variable is the column factor.  But you could have 1 instead.  (Then in the 3 variable you have just 1, 2, 3, so you only allow 1 column factor, and the other 2 are row factors.)  Maybe you could allow 0 in each case, so there are then no column factors - just rows?  Maybe you could do better?

Can you also add the idea from @lily, so it can cope with multiple columns?
Then, before merging it will be good to  also nail the numeric by factor, which is also summaries.

**Comment by Ivanluv** [comment]
Am still working on adding the functionality of displaying more than one frequency tables

**Comment by rdstern** [comment]
@Ivanluv I am hoping that this will, at least partially be in the forthcoming update. By partially:
a)  I am hoping for the 2 variable categorical by categorical with all the options you have (percents, margins), and with multiple 2-way tables that I think you are working on now.
b) Same for Multiple numeric by categorical, which I hope will then be easy.  It is multiple one-way tables.
c) There will be a third button, probably disabled for the 3-way summaries.

I suggest enabling the third button is a separate pull request, and hopefully you will be working on that, and can then continue with that, jointly with @lilyclements when she visits, following the new release.

(And just a pointer that I hope there will be time, during her visit, to progress on conceptualising and starting a new tables dialogue in `Climatic > Check Data > Display Daily Data > Table`.  There is a current "system" which needs replacement.  We have been waiting for a long time, for our own tabulation system to make that replacement.)

**Comment by Ivanluv** [comment]
@rdstern  the changes so far fix the designer changes you suggested and having the display of multiple frequency tables . @lilyclements  the function you suggested ie `survey <- data_book$get_data_frame("survey")
multiple_receiver_cols <- c("fert", "size")
list_of_tables <- survey %>%
  map(.x = multiple_receiver_cols,
      .f = ~data_book$summary_table(data_name = "survey",
                                    columns_to_summarise = .x,
                                    factor = "village",
                                    summaries = "summary_count")) %>%
  map(.x = .,
      .f = ~ mmtable2::mmtable(data=.x, cells=value) +
        mmtable2::header_top_left(variable='summary-variable') +
        mmtable2::header_left_top('village'))
names(list_of_tables) <- multiple_receiver_cols
list_of_tables` 
works but am getting an error when I try adding percentages in the `summary_table` function eg
image
image
image

**Comment by lilyclements** [comment]
@Ivanluv the error you found is a good one and is fixed in PR #7364

If you are interested in the bug:
We have two objects: the summary and the variable. These two objects get merged by a "_".
Sometimes, if we run `treat_columns_as_factors = FALSE`, we wish to split the object at a "_".

However, when the summary already contains a "_" the system gets confused.
_For example, for the summary `perc_count` and variable `fert_col` we would create `perc_count_fert_col`.
When we want to separate, we would usually separate the name by "_". But in this case, we would have `perc` and `count_fert_col`, not `perc_count` and `fert_col`._
As a result, we introduced `-` to specify which to separate by, knowing that none of the summaries had a `-` in their name, and that it would split by the first `-`.
_For example, for the summary `perc_count` and variable `fert_col` we would create `perc_count-fert_col`.
When we want to separate, we would split the name by "-" and get `perc_count` and `fert_col`._

Here comes your bug! In the example you found, it was then creating a variable with a `-` in the name, and R does not allow that. 

E.g., try in R 
`Hi-x <- 100 # does not work`
`Hi_x <- 100 # does work`

This has now been solved and fixed, but a good one nonetheless!

**Comment by Ivanluv** [comment]
@rdstern could you have a look at this

**Comment by rdstern** [comment]
I start with the cosmetic aspects:
a) the title is still Describe Two Variables maybe this can't change till the translation database, but please confirm it will then be Summarise Two/Three Variables?
b) In the middle button change Customise to Two Variables and add the s to Three Variables
c) On the left hand side change categoric to categorical
d) As this is 2 variables don't have the Numeric Variable (Optional) Control - that's for 3 variables.

ok, then please check the working of the rest with different options yourself first.  I couldn't get the simplest to work properly.  So with the usual survey  I tried:
image

I get this sort of thing:
image

This is a poorly labelled one-way table.  Where are the different levels of variety for example?  I didn't realise this was so difficult?  I notice also that each table is on a separate tab in the browser.  Maybe we have to live with that for now?

**Comment by lilyclements** [comment]
@Ivanluv as a very smal point from viewing Rogers screenshot, “customized” is currently spelt with US English. I believe in the software we’ve been using UK England, so for consistency can this change to “customise”?

**Comment by lilyclements** [comment]
> Where are the different levels of variety for example?

@Ivanluv @rdstern this can be added as a stubhead or a caption, but I don’t think mmtable2 offers this as a new cell on the left hand side (that being said I’m sure I can write some code that can do that, depending how high of a priority it is).

**Review by lilyclements — COMMENTED** [review]
@Ivanluv another small point - the “Optional” label has an extra bracket after the colon. Can this extra bracket be removed?

**Comment by Ivanluv** [comment]
@lilyclements I have had a look at this code produced by the  frequency table which I think is what @rdstern wants ie
```
`# Code generated by the dialog, Frequency
frequency_table <- data_book$summary_table(data_name="survey", 
factors=c("village","fertgrp"),
 store_results=FALSE, treat_columns_as_factor=FALSE,
 summaries=count_label)


last_table <- (mmtable2::mmtable(data=frequency_table, cells=value) +
 mmtable2::header_top_left(variable=village) +
 mmtable2::header_left_top(variable=fertgrp))
data_book$add_table(table_name="last_table", table=last_table, data_name="survey")
data_book$get_tables(data_name="survey", table_name="last_table")

rm(list=c("last_table", "frequency_table"))`
```
I was wondering whether we could have this piece of code run multiple times depending on the number of  factors selected eg when we have the dialgue as 

image

we could run the code as

```
`# Code generated by the dialog, Frequency
frequency_table <- data_book$summary_table(data_name="survey", 
factors=c("village","variety"),
 store_results=FALSE, treat_columns_as_factor=FALSE,
 summaries=count_label)


last_table <- (mmtable2::mmtable(data=frequency_table, cells=value) +
 mmtable2::header_top_left(variable=village) +
 mmtable2::header_left_top(variable=variety))
data_book$add_table(table_name="last_table", table=last_table, data_name="survey")
data_book$get_tables(data_name="survey", table_name="last_table")

rm(list=c("last_table", "frequency_table"))
`

`# Code generated by the dialog, Frequency
frequency_table1 <- data_book$summary_table(data_name="survey", 
factors=c("fertgrp","variety"),
 store_results=FALSE, treat_columns_as_factor=FALSE,
 summaries=count_label)


last_table1 <- (mmtable2::mmtable(data=frequency_table, cells=value) +
 mmtable2::header_top_left(variable=fertgrp) +
 mmtable2::header_left_top(variable=variety))
data_book$add_table(table_name="last_table", table=last_table, data_name="survey")
data_book$get_tables(data_name="survey", table_name="last_table")

rm(list=c("last_table1", "frequency_table1"))


`
```

**Comment by Ivanluv** [comment]
@lilyclements any suggestions n this?

**Comment by lilyclements** [comment]
@Ivanluv here's the code we worked on - 

```
diamonds <- data_book$get_data_frame(data_name = "diamonds")
multiple_receiver_cols <- c("cut", "clarity")
list_of_tables <- diamonds %>%
  map(.x = multiple_receiver_cols,
      .f = ~select(data_book$summary_table(data_name = "diamonds",
                                           factor = c("color", .x),
                                           summaries = "summary_count"),
                   c(1, by_var = 2, 3, 4)))
names(list_of_tables) <- multiple_receiver_cols
list_of_tables %>%
  purrr::map(.x = .,
             .f = ~mmtable2::mmtable(data = .x,
                                     cells = value) +
               mmtable2::header_top_left(variable = "by_var") +
               mmtable2::header_left_top(variable = "color"))
```

**Comment by Ivanluv** [comment]
@rdstern @lilyclements could you test this

**Comment by lilyclements** [comment]
@Ivanluv Am I right in thinking that for a three-variable summary, we would have two factor variables and the third is either numeric or factor?

I've put the R code below for this where

* `multiple_receiver_cols` takes the variables in the multiple receiver to run the analysis for each of them
* `color` is the factor variable in one receiver
* `cut` is the factor variable in the other receiver

There are very minor R changes to the code above (add another variable into `factor =`, move the `by_var` in the select, add the new factor in the mmtable2).

```
# script to test code
diamonds <- data_book$get_data_frame(data_name = "diamonds")
multiple_receiver_cols <- c("clarity")
list_of_tables <- diamonds %>%
  map(.x = multiple_receiver_cols,
      .f = ~select(data_book$summary_table(data_name = "diamonds",
                                           factor = c("color", "cut", .x),
                                           summaries = "summary_count"),
                   c(1, 2, by_var = 3, 4, 5)))
names(list_of_tables) <- multiple_receiver_cols
list_of_tables %>%
  purrr::map(.x = .,
             .f = ~mmtable2::mmtable(data = .x,
                                     cells = value) +
               mmtable2::header_top_left(variable = "by_var") +
               mmtable2::header_left_top(variable = "color") +
               mmtable2::header_left_top(variable = "cut"))
```

**Comment by rdstern** [comment]
@lilyclements that's right.  
a) If the multiple receiver is factors and the second is also a factor, and the third is numeric, then there will be multiple 2 (ish) summary tables - with a separate summary table for each factor in the multiple receiver.  There may be multiple variables and could then be options, but the default is probably to have them all in each table.
b) If everything are factors then these are multiple frequency tables.  I assume 3-way frequency tables with again a separate one for each factor in the multiple receiver.
c) In 2 way there is still the multiple numeric, by a factor to update to the new system.  I hope they will also generalise to the 3-way, but there may be combinations we omit for now?

**Comment by lilyclements** [comment]
@Ivanluv is it noted down or added to this PR for the `skimr` package to be installed?

**Review by rdstern — APPROVED** [review]
@lilyclements and @Ivanluv another important step!

**Comment by lloyddewit** [comment]
@Ivanluv Please could you resolve the merge conflict? thanks

File: instat/dlgDescribeTwoVariable.vb [hunk 1]
```diff
@@ -248,4 +482,129 @@ Public Class dlgDescribeTwoVariable
...
+        If rdoCustomize.Checked Then
+            If strFirstVariablesType = "categorical" AndAlso strSecondVariableType = "categorical" Then
+                grpDisplay.Visible = True
+                grpFrequency.Visible = True
+                ucrChkDisplayMargins.Visible = True
+                If ucrChkDisplayMargins.Checked Then
+                    ucrInputMarginName.Visible = True
+                Else
+                    ucrInputMarginName.Visible = False
+                End If
```
Comment (lloyddewit) [inline instat/dlgDescribeTwoVariable.vb:537]:
```suggestion
                ucrInputMarginName.Visible = ucrChkDisplayMargins.Checked
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@Ivanluv Thanks, loogs good.
Just a few small comments

File: instat/dlgDescribeTwoVariable.vb (hunk 1 above)
Comment (Ivanluv) [reply instat/dlgDescribeTwoVariable.vb:537]: fixed in PR #7442

File: instat/dlgDescribeTwoVariable.vb [hunk 2]
```diff
@@ -151,93 +358,120 @@ Public Class dlgDescribeTwoVariable
...
+            Dim iPosition As Integer = 0
+            clsCombineMultipleColumnsFunction.ClearParameters()
+            clsCombineFactorsFunction.ClearParameters()
+
+            For Each strColumn In ucrReceiverFirstVars.GetVariableNamesList()
+                clsCombineMultipleColumnsFunction.AddParameter(strColumn, strColumn, bIncludeArgumentName:=False, iPosition:=iPosition)
+                If ucrReceiverSecondVar.GetVariableNames <> strColumn Then
+                    clsCombineFactorsFunction.AddParameter(strColumn, strColumn, bIncludeArgumentName:=False, iPosition:=iPosition)
+                End If
+            Next
```
[the remark this answers is not in the document]
Comment (Ivanluv) [reply instat/dlgDescribeTwoVariable.vb:466]: Yes, thank you for the reminder. Fixed in PR #7442

File: instat/dlgDescribeTwoVariable.vb [hunk 3]
```diff
@@ -20,8 +20,19 @@ Public Class dlgDescribeTwoVariable
     Private bReset As Boolean = True
     Private bResetSubdialog As Boolean = False
     Public strFirstVariablesType, strSecondVariableType As String
-    Public clsGetDataType, clsGetSecondDataType, clsRCorrelation, clsRCustomSummary, clsConcFunction, clsRAnova, clsRFreqTables As New RFunction
-    Private clsSummariesList As New RFunction
+    Public clsGetDataType, clsGetSecondDataType, clsRCorrelation, clsRCustomSummary,
+           clsCombineFunction, clsRAnova, clsFrequencyTables, clsSkimrFunction, clsSummariesList,
```
[the remark this answers is not in the document]
Comment (Ivanluv) [reply instat/dlgDescribeTwoVariable.vb:24]: Fixed through PR #7442

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 4]
```diff
@@ -1401,67 +1400,88 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
...
+          }else {
+              for (col in 1:ncol(summary_margins)) {
           # TODO: if the colname is the same as a factor, then do nothing
           colnames(summary_margins)[col] <- sub("_value", "_all", colnames(summary_margins)[col])
-        }
-        summary_margins <- summary_margins %>%
-          tidyr::pivot_longer(cols = !factors, names_to = "summary-variable", values_to = "value")
+              }
+              summary_margins <- summary_margins %>%
+              tidyr::pivot_longer(cols = !factors, names_to = "summary-variable", values_to = "value")
```
[the remark this answers is not in the document]
Comment (Ivanluv) [reply instat/static/InstatObject/R/Backend_Components/summary_functions.R]: Fixed in PR #7442

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 5]
```diff
@@ -1401,67 +1400,88 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
...
           tidyr::pivot_longer(cols = !factors, names_to = "summary", values_to = "value")
       } else {
-        for (col in 1:ncol(summary_margins)) {
+          if (length(summary_margins) == 1) {
+              summary_margins <- data.frame(summary_margins, `summary-variable` = "count", factors = NA)
+              names(summary_margins) <- c("value", "summary-variable", factors)
+          }else {
+              for (col in 1:ncol(summary_margins)) {
           # TODO: if the colname is the same as a factor, then do nothing
           colnames(summary_margins)[col] <- sub("_value", "_all", colnames(summary_margins)[col])
```
[the remark this answers is not in the document]
Comment (Ivanluv) [reply instat/static/InstatObject/R/Backend_Components/summary_functions.R:1462]: Fixed in PR #7442

File: instat/static/InstatObject/R/Backend_Components/summary_functions.R [hunk 6]
```diff
@@ -1401,67 +1400,88 @@ DataBook$set("public", "summary_table", function(data_name, columns_to_summarise
...
-        data_book$delete_dataframes(data_names = "summary_margins_df")
+
+      if (is.null(columns_to_summarise)){
+        for (facts in power_sets_summary) {
+          if (length(facts) == 0) facts <- c()
+          summary_margins_df <- data_book$get_data_frame(data_name = data_name) %>%
+            dplyr::select(c(factors))
+          data_book$import_data(data_tables = list(summary_margins_df = summary_margins_df))
+          summary_margins[[length(summary_margins) + 1]] <- data_book$calculate_summary(data_name = "summary_margins_df", columns_to_summarise = NULL, summaries = summaries, factors = facts, store_results = FALSE, drop = drop, na.rm = na.rm, return_output = TRUE, weights = weights, result_names = result_names, percentage_type = percentage_type, perc_total_columns = perc_total_columns, perc_total_factors = perc_total_factors, perc_total_filter = perc_total_filter, perc_decimal = perc_decimal, margin_name = margin_name, additional_filter = additional_filter, perc_return_all = FALSE, ...)
+          data_book$delete_dataframes(data_names = "summary_margins_df")
```
[the remark this answers is not in the document]
Comment (Ivanluv) [reply instat/static/InstatObject/R/Backend_Components/summary_functions.R:1433]: @lilyclements could you have a look at the code

**Comment by lloyddewit** [comment]
Replaced by PR #7442

**Pull request closed without merging**